### PR TITLE
Resume tee-worker

### LINKED ISSUE #997: Can't reconnect to the parentchain once disconnected
I assume this is a different problem other than litentry/litentry-parachain#999, once the worker node lost connection with the parentchain (e.g. due to an unstable network), it can't reconnect.

Maybe related: https://github.com/integritee-network/worker/issues/813

### LINKED ISSUE #999: Can not resume worker node
As title

## Commands
 - Run worker1 for the first time 
```bash
./integritee-service --clean-reset --mu-ra-external-address localhost --mu-ra-port 3443 \
--untrusted-http-port 4545 --ws-external --trusted-external-address wss://localhost \
--trusted-worker-port 2000 --untrusted-external-address ws://localhost \
--untrusted-worker-port 3000 --node-url ws://host.docker.internal \
--node-port 9946 run --skip-ra --dev
```

 - Run worker2 for the first time (wait for worker1 initialize successfully)
```bash
dockerize -wait http://localhost:4545/is_initialized -timeout 600s

./integritee-service --clean-reset --mu-ra-external-address localhost --mu-ra-port 3444 \
--untrusted-http-port 4546 --ws-external --trusted-external-address wss://localhost \
--trusted-worker-port 2001 --untrusted-external-address ws://localhost \
--untrusted-worker-port 3001 --node-url ws://host.docker.internal \
--node-port 9946 run --skip-ra --request-state
```

 - Resume worker1
```bash
./integritee-service --mu-ra-external-address localhost --mu-ra-port 3443 \
--untrusted-http-port 4545 --ws-external --trusted-external-address wss://localhost \
--trusted-worker-port 2000 --untrusted-external-address ws://localhost \ 
--untrusted-worker-port 3000 --node-url ws://host.docker.internal \
--node-port 9946 run --skip-ra
```

- Resume worker2
```bash
./integritee-service --mu-ra-external-address localhost --mu-ra-port 3444 \
--untrusted-http-port 4546 --ws-external --trusted-external-address wss://localhost \
--trusted-worker-port 2001 --untrusted-external-address ws://localhost \
--untrusted-worker-port 3001 --node-url ws://host.docker.internal \
--node-port 9946 run --skip-ra
```

## Scenarios

Currently we have tried 3 different ways to resume the stopped node, but none of them worked.

### 1. One worker
steps to reproduce:
1. Launch parachain
2. Run worker1 for the first time. Please see above command.
3. After worker1 initialize successfully, stop it.
4. Resume worker1. Please see above command.

### 2. Two workers - Stop worker1(Primary worker?)
steps to reproduce:
1. Launch parachain
2. Run worker1 for the first time. Please see above command.
3. After worker1 initialize successfully, run worker2 for the first time.
4. After worker2 initialize successfully, stop worker1.
5. Resume worker1. Please see above command.

### 3. Two workers - Stop worker2
steps to reproduce:
1. Launch parachain
2. Run worker1 for the first time. Please see above command.
3. After worker1 initialize successfully, run worker2 for the first time.
4. After worker2 initialize successfully, stop worker2.
5. Resume worker2. Please see above command.

### PR DESCRIPTION
**Pull request by ziming-zung**
resolves: #999
~~resolves: #997~~

This PR:

- After stopping, the tee-worker can resumes syncing
- tee-worker attempt to reconnect the parachain  once disconnected

For #999 
1. We should change the `read_or_init_parachain_validator` function in **core/parentchain/light-client/src/io.rs**.
In this function, the state is cleared every time, so we should rewrite the state after init.
https://github.com/litentry/litentry-parachain/blob/09784f1d74f3882f22062579178ea3041e8daeed/tee-worker/core/parentchain/light-client/src/io.rs#L79-L84

2. Enclave will be registered each time  when tee-worker run. So we shoud check if enclave is already registered each time.
https://github.com/litentry/litentry-parachain/blob/09784f1d74f3882f22062579178ea3041e8daeed/tee-worker/service/src/main.rs#L444-L446

**Strongly recommended to manually execute the [command](https://github.com/litentry/litentry-parachain/issues/999#issue-1467762720) to confirm**

For #997 
substrate-api-client uses library [ws](https://github.com/housleyjk/ws-rs), which is difficult to implement reconnection mechanism. So I decided to use [tungstenite](https://github.com/snapview/tungstenite-rs) instead of [ws](https://github.com/housleyjk/ws-rs). 
To check implementation, see at: https://github.com/litentry/substrate-api-client/tree/new-tungstenite-polkadot-v0.9.29

### CODE REVIEW

**Review by Kailai-Wang — APPROVED** [review]
Good job locating the problem.

While there might still be some small problems with reconnection, I'm thinking if we can add tests for it? These tests are actually very important.

We'd need to test:

- parentchain disconnected and reconnected, the workers (both primary and non-primary validateers can still continue 
- primary validateers can stop and resume
- non-primary validateers can stop and resume

After we have the tests we can submit a PR to upstream

**Comment by Kailai-Wang** [comment]
Looks like still some CI issues?

File: tee-worker/service/src/main.rs [hunk 1]
```diff
@@ -441,15 +444,59 @@ fn start_worker<E, T, D, InitializationHandler, WorkerModeProvider>(
...
+					hex::decode(bytes).unwrap()
+				} else {
+					hex::decode(key.as_bytes()).unwrap()
+				};
+				match node_api.get_storage_by_key_hash::<TeerexEnclave<AccountId32, Vec<u8>>>(
+					StorageKey(key.clone()),
+					None,
+				) {
+					Ok(Some(value)) => {
+						if value.mr_enclave.to_vec() == mrenclave && value.url == trusted_url {
```
Comment (Kailai-Wang) [inline tee-worker/service/src/main.rs:533]: It's almost guaranteed that we'll fall into the `!found` branch if we are worker-2, as the `trusted_url` is different. I don't understand why we still need to register the enclave in this case, but it's the same logic as the original code

**Comment by Kailai-Wang** [comment]
Update: we might have some problems in the `tungstenite-client` version of substrate-api-client -- let's try to update it and re-test it

**Comment by ziming-zung** [comment]
> Update: we might have some problems in the `tungstenite-client` version of substrate-api-client -- let's try to update it and re-test it

I had updated the `tungstenite-client`

**Comment by Kailai-Wang** [comment]
CI looks good!

I think we can merge this one and submit a PR to `substrate-api-client`

**Comment by Kailai-Wang** [comment]
Maybe let's merge this PR and write tests in another PR? @zhizming-zhong

**Comment by ziming-zung** [comment]
> Maybe let's merge this PR and write tests in another PR? @zhizming-zhong

Of course, after conflicts are resolved and the CI is passed, this PR will be merged into dev branch

**Comment by ziming-zung** [comment]
notice !!!
I reset the branch, and this PR only resolves  #999.
may be need to wait for  the PR  [`Make api no_std compatible`](https://github.com/scs/substrate-api-client/pull/377) to be completed before solving issue #997

**Review by Kailai-Wang — APPROVED** [review]
Looks good, thanks

**Pull request merged**